### ISSUE TITLE
Requirements for taints, tolerations, dedicated nodes to graduate to Beta and then v1

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#25320]
We'd like to graduate taints, tolerations, and dedicated nodes to Beta in 1.4. (Assuming we get them into 1.3 as alpha.)

Taints/tolerations to Beta
- [x] Add forgiveness to Toleration
- [x] make taints unique by `<key, effect>` on a node #29362
- [ ] Change annotation key from alpha to beta
- [ ] Maybe make Toleration key wildcard-able
- [ ] Move to GeneralPredicates
- [ ] Proper defaulting (when it's a field)
- [ ] Implement NoScheduleNoAdmit and NoExecute
  (previously NoScheduleNoAdmitNoExecute see #28082)
- [ ] Prevent users from modifying taints/tolerations (see e.g. #17549)
- [ ] Ensure that a new node does not become "Ready" until it has been configured with its taints. One way to do this is to have an admission controller that adds the taint whenever a Node object is created.
- [ ] Before allowing users to use it, be sure we won't need to roll back to binary version that doesn't support it

Taints/tolerations to v1
- [ ] Move from annotations on Pod and Node to fields of PodSpec and NodeSpec

Dedicated nodes to Beta
- [ ] More sophisticated policy for adding tolerations, and maybe easier way for adding taints (special-cased to dedicated nodes use case)

cc/ @kevin-wangzefeng @bgrant0607 

ref/ #24134 #17190 #18263

**Comment by resouer** [comment]
Working on node side to support kubelet enforced feature of taints/toleration

**Comment by resouer** [comment]
@davidopp Cloud you explain why add forgiveness to Toleration? If we don't want to tolerate some taints, just remove toleration is not enough?

**Comment by davidopp** [comment]
The canonical example for forgiveness is: let's say we represent node problems using taints (which we actually do plan to do). So for example Then a pod can use toleration to represent how long it wants to stay bound to a nod before it should be evicted.  You want to add the taint immediately when the node starts having the problem, because some pods might want to be evicted right away. But the ones that don't want to be evicted right away, can use toleration to specify how long they want to stay bound. (Of course this assumes the TaintEffect is TaintEffectNoScheduleNoAdmitNoExecute.)

BTW @kevin-wangzefeng is going to work on adding forgiveness to Toleration.

**Comment by resouer** [comment]
Thanks, its much clearer.

**Comment by hongchaodeng** [comment]
@davidopp 

I would to help implement some tasks for v1.4. Can your mark tasks being taken by somebody already? Or can folks who is taking over mark themselves here?

BTW, what's "proper defaulting"?

**Comment by davidopp** [comment]
Yes, sorry, to clarify, we haven't done any planning for 1.4, and we intend to do it jointly with the community when it happens. But @kevin-wangzefeng implemented taints and tolerations (#24134) and had already told me he was planning to add forgiveness since it was part of the original design.

**Comment by resouer** [comment]
@hongchaodeng Except for `forgiveness` and what I was working on e.g. NoScheduleNoAdmit and NoScheduleNoAdmitNoExecute, it seems other todos are all available. Would be excited to see you pick some up.

**Comment by lukaszo** [comment]
Is anyone working on ` Move to GeneralPredicates` ? If no, I can take it.

**Comment by davidopp** [comment]
@lukaszo  I don't think anyone is working on it.

**Comment by kevin-wangzefeng** [comment]
Some updates:
- added an new item **make taints unique by <key, effect> on a node #29362**, to clarify what make a taint unique on a node.
- updated item **Implement NoScheduleNoAdmit and NoExecute (previously NoScheduleNoAdmitNoExecute, see #28082)** `NoScheduleNoAdmitNoExecute` is now to use `NoScheduleNoAdmit` + `NoExecute` instead

**Comment by eparis** [comment]
@davidopp is this list still correct(ish) for getting to beta? Is there any better place to look?

**Comment by davidopp** [comment]
I think this list is more or less accurate. @kevin-wangzefeng is working on some of the items for 1.5. I don't think we will have time to make it Beta in 1.5 though. We could maybe split this up into taints/tolerations and dedicated nodes, and consider the latter a separate feature; that would allow us to remove the security-related items from the list (like ensuring Kubelet comes up with the right taints when it joins, and preventing unprivileged users from modifying taints/toleations). Also we definitely don't need NoScheduleNoAdmit for Beta, as it can be added in a backward-compatible way.

**Comment by davidopp** [comment]
Figure out what to do with NodeStatus.NodeSpec "unschedulable" field

**Comment by aveshagarwal** [comment]
@kevin-wangzefeng @davidopp Could you please share what specific items kevin is working on for 1.5, any links or anything for reference? Also I am happy to help on other items not being worked on by kevin, perhaps to help it move closer to beta.

@derekwaynecarr

**Comment by timothysc** [comment]
@aveshagarwal please coordinate via the @kubernetes/sig-scheduling to ensure folks aren't overlapping.   xref: http://goo.gl/CsXJLu

**Comment by aveshagarwal** [comment]
I am trying to summarize the tasks here to capture the current state based on what I have found as its not clear what is being worked on, is already done or is yet to do. As @davidopp suggested to separate taints/tolerations and dedicated nodes features in one of his previous comments, I am updating the list for taints/tolerations and dedicated nodes features to graduate to beta here (please feel free to correct me, and also update as needed ):

Taints/Toleration to Beta:

To-do:
- [ ] Change annotation key from alpha to beta
- [ ] Maybe make Toleration key wildcard-able
- [ ] Before allowing users to use it, be sure we won't need to roll back to binary version that doesn't support it

Being worked on/or done:
- [ ] Add forgiveness to Toleration  (https://github.com/kubernetes/kubernetes/pull/30191)
- [ ] Implement NoExecute (https://github.com/kubernetes/kubernetes/pull/30191)
- [ ] Move to GeneralPredicates  (Ref: https://github.com/kubernetes/kubernetes/pull/31907, https://github.com/kubernetes/kubernetes/pull/31020)

Taints/tolerations to v1:

- [ ]  Implement NoScheduleNoAdmit
- [ ]  Move from annotations on Pod and Node to fields of PodSpec and NodeSpec
- [ ]  Proper defaulting (when it's a field)


Dedicated nodes to Beta

- [ ]  Prevent users from modifying taints/tolerations (see e.g. #17549) (Ref https://github.com/kubernetes/kubernetes/pull/27330)
- [ ]  Ensure that a new node does not become "Ready" until it has been configured with its taints. One way to do this is to have an admission controller that adds the taint whenever a Node object is created.
- [ ]  More sophisticated policy for adding tolerations, and maybe easier way for adding taints (special-cased to dedicated nodes use case)

We can discuss more about it in today's sig-scheduling meeting.

**Comment by aveshagarwal** [comment]
@kubernetes/huawei

**Comment by aveshagarwal** [comment]
@davidopp @kevin-wangzefeng As discussed in the yesterday's sig-scheduling meeting, would like to have feedback on my last comment or in general.

**Comment by timothysc** [comment]
@aveshagarwal things to keep in mind re(alpha-beta) : https://github.com/kubernetes/kubernetes/issues/30819#issuecomment-249877334

**Comment by davidopp** [comment]
Sorry for the delay. I think my list was way too long. Those are all features we want eventually but I don't think they're all needed for 1.5.

The remaining things I think we need for 1.5 are
* add forgiveness to Toleration
* Implement NoScheduleNoAdmit and NoExecute
(previously NoScheduleNoAdmitNoExecute see #28082)

@kevin-wangzefeng is working on these and said he should have a PR out soon.

As for graduating it to Beta, we should do that in a followup PR. I hope we can get it done for 1.5. I think we need to implement "make toleration key wildcardable" before (or simultaneous with) graduating to Beta because I'm not sure we can add it later in a backward compatible way. (imagine someone creates an alpha toleration with key X and then we decide that in beta key X means wildcard; then when they upgrade from alpha to beta their toleration suddenly means something new)

There is discussion ongoing in #30819 about how to graduate things to Beta in general.

We can ignore dedicated notes and just consider it a separate feature to be implemented in the future.

**Comment by aveshagarwal** [comment]
I can take a stab at "make toleration key wildcardable" implementation if no one is working on it.  Here is my understanding. 

When a toleration key is wild card:
Case 1: Operator=Equal, Value=v, Effect=one of NoSchedule, PreferNoSchedule, NoExecute, NoScheduleNoAdmit

It means that a pod with the above toleration could be scheduled to any node that has a taint with value v (irregardless of key) according to the effects.

Case 2: Operator=Exits, Value=v, Effect=one of NoSchedule, PreferNoSchedule, NoExecute, NoScheduleNoAdmit

In this case, the value is wildcard (As Operator=Exits) too. As both key and value are wild cards, it seems equivalent to no toleration.

**Comment by davidopp** [comment]
I think @kevin-wangzefeng may have started working on it, so you should check with him.

What I proposed was that empty toleration key (empty string) means "match all keys"
I think it would only make sense to use it with Exists operator (so it means "match all values and all keys"), so we would check that in validation.

**Comment by aveshagarwal** [comment]
@davidopp thanks, its ok with me if @kevin-wangzefeng is working on it, so I will leave it to him.

Is someone working on "NoScheduleNoAdmit" implementation too, as I haven't seen it any PR? In case no, I would be happy to help with it.

**Comment by kevin-wangzefeng** [comment]
@aveshagarwal I think it's #34400 (previously #26501), but I didn't get time to take a look yet.

**Comment by aveshagarwal** [comment]
@kevin-wangzefeng Yes https://github.com/kubernetes/kubernetes/pull/34400 implements NoScheduleNoAdmit.

**Comment by aveshagarwal** [comment]
Is this the expectation with NoScheduleNoAdmit?

a) NoSchedule part of NoScheduleNoAdmit is exactly similar to existing NoSchedule and is for pods going via scheduler. 

b) NoAdmit part of NoScheduleNoAdmit is for static pods (created directly by kubelet, i..e the pods not going through the scheduler).

**Comment by davidopp** [comment]
Unfortunately we're not going to have time to get #34400 into 1.5.

But as for what you described @aveshagarwal, it is correct except I'm not sure the term "static pods" is correct (or maybe I just don't understand the definition you're using). But as you said, it is for pods that are submitted directly to Kubelet, e.g. by creating a pod with NodeName already filled in.

**Comment by aveshagarwal** [comment]
@davidopp I just picked the word static pod from here: http://kubernetes.io/docs/admin/static-pods/

**Comment by davidopp** [comment]
Yeah I'm not sure whether NoAdmit would affect static pods. The intended use case was a regular pod that sets NodeName directly. Static pods are a bit different as they are created from a manifest file (and in theory by HTTP but nobody does that).

**Comment by aveshagarwal** [comment]
I have copied the list from https://github.com/kubernetes/kubernetes/issues/35518 here for taints/tolerations too :

1. Remove the existing implementation in annotations.
2. Implement existing stuff in api fields (xref: https://github.com/kubernetes/kubernetes/issues/34508).
3. Port tests.

As we discussed in yesterday's sig-sched meeting, I will start looking into it once the existing PRs are in better/working shape probably for 1.6 time frame. 

Once functionality has reached parity with the annotation implement, promote from alpha->beta.

@kubernetes/sig-scheduling  thoughts?

**Comment by timothysc** [comment]
@aveshagarwal I'd like to sync sometime on your test plan(s) around this.  Right now everything is lite at best.  

but +1 to the plan.

**Comment by Quentin-M** [comment]
We have an use-case where, we'd like a pod managed by a DaemonSet to detect when the underlying node is being put in maintenance mode. Currently, this means _kubectl cordon_ or _kubectl drain_ are ran and set the _NodeStatus.NodeSpec.Unschedulable_ field. But because the pod is part of a DaemonSet, it is not destroyed and there is no way to differentiate between the two operations using the API. However, it actually has quite some importance to us because in the case where the node is drained, it shows the operator's intent to shutdown/restart the node (or realize some other heavy duty task), in which case we want to gracefully migrate our workload/tasks/resources to another working pod/node. In the other hand, if the node is simply cordoned, the risk is limited (pods restarting is fine) and it is not worth migrating.

Unfortunately, as mentioned above, we currently can't tell and therefore, we are forced to migrate everything as soon as the node is set as unschedulable, even if it's a very short maintenance/outage.

It seems that the _NoScheduleNoAdmit_ and _NoExecute_ taints fit the use-case. However, I didn't clearly see any mention about how _kubectl cordon_ and _kubectl drain_ will be reworked to adopt these. Intuitively, I'd say that _kubectl cordon_ would set _NoScheduleNoAdmit_ while _kubectl drain_ would set _NoScheduleNoAdmit_ + _NoExecute_, without actively deleting any pods but instead letting the schedule do it. Tolerations and Forgiveness could then be used to make my DaemonSet's pods stop and execute the necessary migration steps using the _preStop_ hook - While at the same time, DaemonSets such as kube-proxy would tolerate these taints and keep running indefinitely.

Does it make sense? Am I missing anything? I believe these are planned for 1.6?

**Comment by aveshagarwal** [comment]
@Quentin-M yes the current plan for NoScheduleNoAdmit and NoExecute implementation is 1.6. Here are related PRs: https://github.com/kubernetes/kubernetes/pull/34825 , https://github.com/kubernetes/kubernetes/pull/34400 .

**Comment by davidopp** [comment]
Each pod in a DaemonSet is supposed to be tied to the node it is running on. There shouldn't be any need to migrate anything when the machine goes down. So it sounds like you are using DaemonSet in an unexpected way.

But anyway, what you and @aveshagarwal is indeed the plan; we will switch cordon/uncordon/drain to use taints, and then DaemonSet pods can decide for themselves whether they want to be deleted when these actions happen. We were hoping to get this into 1.5 but it didn't make it, but it should be in 1.6. #29178 covers the "represent unschedulable as a taint" part; i don't think we have an issue filed for making kubectl drain use taints but we can just leave your comment here about it as a reminder.

**Comment by davidopp** [comment]
Actually I don't think we can use taints for kubectl drain, because we want to respect PodDisruptionBudget when doing the evictions for a drain, and eviction due to NoExecute taint presumably won't. I could imagine some scheme where the NoExecute taint could specify whether the eviction should respect PDB (go through pod /eviction subresource) or not, but this seems complicated. So I don't think eviction for taint will respect PDB. If we find some important use cases where we do need it, then we could do something like what I described, but it's complication we should avoid unless it's truly necessary.

**Comment by davidopp** [comment]
We should consider whether to add taint/toleration check to GeneralPredicates (so Kubelet checks them) when moving to Beta.

**Comment by davidopp** [comment]
We need to write documentation. The documentation should emphasize that this is not really intended to be used by users directly, but by higher-level tools that would apply the appropriate taints and configure an admission controller to add the appropriate tolerations. We should also mention that you need to pair a toleration with a node affinity in some scenarios.

**Comment by davidopp** [comment]
Once we think the work is done, we need to verify that no instances remain in the codebase of the strings "TolerationsAnnotationKey", "api.TolerationsAnnotationKey", "scheduler.alpha.kubernetes.io/tolerations", "TaintsAnnotationKey", "api.TaintsAnnotationKey", "scheduler.alpha.kubernetes.io/taints"

**Comment by aveshagarwal** [comment]
@davidopp Yeah sure. Though this PR https://github.com/kubernetes/kubernetes/pull/38957 has already removed them. So I will update this PR after forgiveness related work is merged.

**Comment by davidopp** [comment]
ref/ #41172 #39469 #39914 #40355 #41014 #41068 #41073 #41133

**Comment by davidopp** [comment]
Remaining work to move taints & tolerations to Beta:
- [ ] Make cluster autoscaler 1.6 use field-based taints and remove annotation-based taints (with taint key “ToBeDeletedByClusterAutoscaler”): assigned to @mwielgus 
- [ ] Make rescheduler 1.6 use field-based taints and remove annotation-based taints (with taint key “CriticalAddonsOnly”) and use field-based tolerations and remove annotation-based tolerations (for taint key “CriticalAddonsOnly”): assigned to @piosz 
- [ ] Write release notes for rescheduler users: assigned to @piosz 
- [ ] PR to move taints and tolerations from annotations to fields (#38957) needs to be completed and merged **after** following two bullets: assigned to @aveshagarwal 
- [ ] Address comments on taint controller review: assigned to @gmarek 
- [ ] PR to (optionally) use taints to trigger eviction of pods due to node problems, instead of direct eviction by NodeController (#41133), needs to be completed and merged: assigned to @gmarek 

Technically the last bullet is not necessary for moving these features to Beta but it is closely related and #41133 is a lot of churn so it will be easier to get that in if we wait with #38957 until after it goes in.

Upgrading to 1.6 will break existing users of these features and they will have to switch to fields, but this within the bounds of our policy for alpha features. Information about upgrading has been sent to kubernetes-dev ([here](https://groups.google.com/d/msg/kubernetes-dev/-fwfHQtm9GE/H1TU4yqiBAAJ)) and kubernetes-users ([here](https://groups.google.com/d/msg/kubernetes-users/xf-VlRlAm2A/Th76aF-lAwAJ)) and will also appear in the release notes.

**Comment by resouer** [comment]
@davidopp I think we can now continue the implementation `TaintEffectNoScheduleNoAdmit` in #34400, right? Since we are aiming it at 1.6 ref https://github.com/kubernetes/kubernetes/issues/25320#issuecomment-261571143

If so I am going to update #34400 very quickly.

**Comment by davidopp** [comment]
@resouer I don't think we should do NoScheduleNoAdmit for 1.6. It is too much changing in one release and I have not heard any strong requests for it from customers (of Google or any of our partners). It is definitely useful for some scenarios (e.g. "protect" the kubelet when pod is not scheduled by default scheduler) but I think the risk outweighs the benefit for putting it into 1.6. I think it would be better to target 1.7.

If there has been strong request from customers you are aware of, please let me know and we could reconsider. But I am already nervous about the amount of scheduling stuff that we are enabling in 1.6.

**Comment by resouer** [comment]
@davidopp That's fair enough, just read this comment and why I'm asking:
```
@Quentin-M yes the current plan for NoScheduleNoAdmit and NoExecute implementation is 1.6. Here are related PRs: #34825 , #34400 .
```

Let's keep our plan as it is.

**Comment by davidopp** [comment]
Yeah, I am not sure the background for that comment. Maybe at one point we said we were going to do it for 1.6 (I can't remember), but I think that in light of all of the other changes it would be safer to delay it to 1.7. Sorry about that.

**Comment by gmarek** [comment]
There's a subbullet for the last checkbox:
- write a admission controller that'll set default tolerations for NC taints to all Pods that doesn't have them.

**Comment by davidopp** [comment]
Oh yes -- thanks for remembering that! Will you have time to do that? Or do you think @kevin-wangzefeng could do it? (He has written an admission controller before.)

**Comment by gmarek** [comment]
I asked @kevin-wangzefeng on some PR, but he didn't respond. If he won't do it today or tomorrow I'll need to write it.

**Comment by yastij** [comment]
@gmarek @davidopp - As #57690 started, moving to Taints requires thinking about `NodeStatus.NodeSpec.Unschedulable` and if we still consider to transform it into a Taint. 

For example service controller relies on this field to list node based on their conditions (and #57690 it will support listing using taints). 

If we still want to transform it into a taint, I might be able to do so, as it might be needed for #57690

**Comment by yastij** [comment]
ping @gmarek 

also cc @etiennecoutaud

**Comment by gmarek** [comment]
IIRC `Unschedulable` was supposed to be deprecated in favor of Taints. Desired usecase for it was to allow cluster operator to tell the system where it shouldn't schedule Pods, which currently can be done by Taints in much better and more configurable way.

If anything currently depends solely on `Unschedulable` it's a bug and should be migrated to look at NoSchedule/NoExecute Taints as well. @davidopp @bgrant0607

**Comment by yastij** [comment]
@gmarek - some parts still depends on this(kubelet and service_controller and some other parts), It'll be migrated with the nodeConditions.

**Comment by gmarek** [comment]
Yes, we're not planning to stop respecting `unschedulable`, but there should be no place in the system which looks at `unschedulable` only and ignores Taints.

I guess it is, but we let it drift a bit from basic Taints/Tolerations towards Taint/Toleration base features. @yastij - can you file a separate issue for that?

**Comment by bgrant0607** [comment]
cc @kubernetes/sig-scheduling-proposals

**Comment by k82cn** [comment]
/assign

Let me go through current status of Taint/Toleration. Currently, `Unschedulable by Taints` is supported as an alpha feature, `TaintBasedEviction and TaintNodesByCondtion` are alpha; but no impact to Taint/Toleration graduation.

**Comment by tallclair** [comment]
What's the status of this?

I noticed that the `TolerationsAnnotationKey` appears to only be used in validation.

**Comment by pacoxu** [comment]
/assign
If this will be in 1.22 or late releases.

**Comment by sftim** [comment]
I propose we unfreeze this issue. I don't think there's work to do, _and_ it's not a KEP.

**Comment by alculquicondor** [comment]
This issue is too old.

individual items should be worked on via KEPs

/close
